DuplicateTokenEx

zabulus opened the pull request.

zabulus committed:
Add DuplicateTokenEx

vbfox commented on src/AdvApi32.Desktop/AdvApi32+TokenType.cs:
```diff
@@ -0,0 +1,34 @@
+        /// The <see cref="TokenType"/> enumeration contains values that differentiate between a primary token and an
+        /// impersonation token.
+        /// </summary>
+        public enum TokenType
```
This enumeration is named `TOKEN_TYPE` on MSDN : https://msdn.microsoft.com/en-us/library/windows/desktop/aa379633(v=vs.85).aspx

vbfox commented on src/AdvApi32.Desktop/AdvApi32.cs:
```diff
@@ -896,6 +896,38 @@ public static partial class AdvApi32
+        [DllImport(api_ms_win_security_base_l1_2_0, SetLastError = true)]
+        public static extern bool DuplicateTokenEx(
+            SafeObjectHandle hExistingToken,
+            uint dwDesiredAccess,
```
I don't remember the details of discussions about access rights @AArnott should we create a specific `ACCESS_MASK` this time or wait for a decision on unification (#271)

(Knowing that theThe access rights are already present as uint constants in `AdvApi32+TokenAccessRights.cs`)

AArnott replied:
We should use `ACCESS_MASK` for the time being. It will allow conversion to `uint` when necessary but in the meantime at least captures a bit of the strong-typed nature we're going for.

arlm replied:
There are samples for the specific and special flags, if needed, on the inner enumerations.

vbfox commented on src/AdvApi32.Desktop/AdvApi32.cs:
```diff
@@ -896,6 +896,38 @@ public static partial class AdvApi32
+        public static extern bool DuplicateTokenEx(
+            SafeObjectHandle hExistingToken,
+            uint dwDesiredAccess,
+            IntPtr lpTokenAttributes,
```
We have the structure, so it should be of type `SECURITY_ATTRIBUTES*`.

AArnott replied:
Yes, with a `[Friendly(In|Optional)]` attribute on it.

zabulus commented:
Squashed commits, after my ugly merge they lost any meaning.
Addressed review issues:
1. TOKEN_TYPE naming
2. ACCESS_MASK
3. SECURITY_ATTRIBUTES (including attribute)
4. Also return BOOL marshalling

AArnott commented:
Thanks for your contribution.
